src/providers/ws.rs

```diff
@@ -0,0 +1,497 @@
+use std::{
+    collections::{
+        btree_map::Entry,
+        BTreeMap,
+    },
+    fmt::Debug,
+};
+
+use async_trait::async_trait;
+use base64::{
+    engine::general_purpose,
+    Engine,
+};
+use ethers::types::H160;
+use futures::{
+    future,
+    select,
+    stream::Fuse,
+    Sink,
+    SinkExt,
+    Stream,
+    StreamExt,
+    TryStreamExt,
+};
+use futures_channel::mpsc;
+use tokio_tungstenite::connect_async;
+use tracing::{
+    debug,
+    error,
+    trace,
+    warn,
+};
+use tungstenite::{
+    client::IntoClientRequest,
+    Message,
+};
+use uuid::Uuid;
+
+use crate::core::{
+    error::{
+        Error,
+        Result,
+    },
+    provider::{
+        EthereumProvider,
+        Provider,
+        StreamResponse,
+        UniswapV2Provider,
+        UniswapV3Provider,
+    },
+    types::{
+        ethereum::BlockHeader,
+        query::QueryOptions,
+        uniswap_v2,
+        uniswap_v3,
+    },
+};
+
+const WS_PATH: &str = "websocket";
+
+type WsMsg = Result<tungstenite::Message, tungstenite::Error>;
+type WsResult = Result<Vec<u8>>;
+type OperationMsg = (Uuid, Operation, mpsc::UnboundedSender<WsResult>);
+
+#[derive(Clone, Debug)]
+pub struct WsProviderImpl {
+    operations: mpsc::UnboundedSender<OperationMsg>,
+}
+
+impl WsProviderImpl {
+    async fn request<T>(&self, operation: Operation) -> StreamResponse<T>
+    where
+        T: serde::de::DeserializeOwned + 'static,
+    {
+        let raw_data_stream = self.raw_request(operation).await?;
+        let raw_data_stream = raw_data_stream.into_async_read();
+
+        let stream = csv_async::AsyncDeserializer::from_reader(raw_data_stream);
+        let records = stream
+            .into_deserialize::<T>()
+            .map(|r| {
+                match r {
+                    Ok(r) => Ok(r),
+                    Err(e) => Err(Error::from(e)),
+                }
+            })
+            .boxed();
+
+        Ok(records)
+    }
+
+    async fn raw_request(
+        &self,
+        operation: Operation,
+    ) -> Result<impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Send> {
+        let (sink, stream) = mpsc::unbounded();
+        let id = Uuid::new_v4();
+
+        self.operations
+            .unbounded_send((id, operation, sink))
+            .map_err(|_| Error::BackendShutDown)?;
+
+        let stream = stream
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
+            .filter_map(|data| {
+                match data {
+                    Ok(data) => {
+                        if data.is_empty() {
+                            return future::ready(None);
+                        }
+                        future::ready(Some(Ok(data)))
+                    }
+                    Err(e) => future::ready(Some(Err(e))),
+                }
```

Comment:
This could be simplified a bit
```suggestion
            .filter_map(|data| async {
                match data {
                    Ok(data) if !data.is_empty() => Some(Ok(data)),
                    Ok(_) => None,
                    Err(e) => Some(Err(e)),
```